server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -484,11 +485,66 @@ void executeRequest(
                 }
             }
         });
+        final SearchSourceBuilder source = original.source();
+        if (shouldOpenPIT(source)) {
+            openPIT(original, searchService.getDefaultKeepAliveInMillis(), listener.delegateFailureAndWrap((delegate, resp) -> {
+                source.pointInTimeBuilder(new PointInTimeBuilder(resp.getPointInTimeId()).setKeepAlive(TimeValue.MINUS_ONE));
+                executeRequest(task, original, new ActionListener<>() {
+                    @Override
+                    public void onResponse(SearchResponse response) {
+                        // we need to close the PIT first so we delay the release of the response to after the closing
+                        response.incRef();
+                        closePIT(original.source().pointInTimeBuilder(), () -> ActionListener.respondAndRelease(listener, response));
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        closePIT(original.source().pointInTimeBuilder(), () -> listener.onFailure(e));
+                    }
+                }, searchPhaseProvider);
+            }));
+        } else {
+            Rewriteable.rewriteAndFetch(
+                original,
+                searchService.getRewriteContext(timeProvider::absoluteStartMillis, resolvedIndices, original.pointInTimeBuilder()),
+                rewriteListener
+            );
+        }
+    }
+
+    /**
+     * Returns true if the provided source needs to open a shared point in time prior to executing the request.
+     */
+    private boolean shouldOpenPIT(SearchSourceBuilder source) {
```

Comment:
could this be made static?

Reply:
https://github.com/elastic/elasticsearch/pull/110482/commits/a370caca0fb026890e6a8a893591658fec6a6fb6